modules/cart/tests/src/FunctionalJavascript/MultipleCartFormsTest.php

```diff
@@ -0,0 +1,216 @@
+
+/**
+ * Tests pages with multiple products rendered with add to cart forms.
+ *
+ * @group commerce
+ */
+class MultipleCartFormsTest extends CartBrowserTestBase {
+
+  use JavascriptTestTrait;
+
+  /**
+   * Modules to enable.
+   *
+   * @var array
+   */
+  public static $modules = [
+    'commerce_cart_test',
+    'commerce_cart_big_pipe',
+  ];
+
+  /**
+   * @var \Drupal\commerce_product\Entity\ProductAttributeValueInterface[]
+   */
+  protected $colorAttributes = [];
+
+  /**
+   * @var \Drupal\commerce_product\Entity\ProductAttributeValueInterface[]
+   */
+  protected $sizeAttributes = [];
+
+  /**
+   * @var \Drupal\commerce_product\Entity\ProductInterface[]
+   */
+  protected $products = [];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+
+    // Delete parent test product.
+    $this->variation->getProduct()->delete();
+    $this->variation->delete();
+
+    /** @var \Drupal\Core\Entity\Entity\EntityFormDisplay $order_item_form_display */
+    $order_item_form_display = EntityFormDisplay::load('commerce_order_item.default.add_to_cart');
+    $order_item_form_display->setComponent('quantity', [
+      'type' => 'number',
+    ]);
+    $order_item_form_display->save();
+
+    $variation_type = ProductVariationType::load('default');
+    $color_attributes = $this->createAttributeSet($variation_type, 'color', [
+      'red' => 'Red',
+      'blue' => 'Blue',
+    ]);
+    $this->colorAttributes = $color_attributes;
+    $size_attributes = $this->createAttributeSet($variation_type, 'size', [
+      'small' => 'Small',
+      'medium' => 'Medium',
+      'large' => 'Large',
+    ]);
+    $this->sizeAttributes = $size_attributes;
+
+    $attribute_values_matrix = [
+      ['red', 'small'],
+      ['red', 'medium'],
+      ['red', 'large'],
+      ['blue', 'small'],
+      ['blue', 'medium'],
+      ['blue', 'large'],
+    ];
+
+    for ($i = 1; $i < 5; $i++) {
+      // Create a product variation.
+      $variations = [];
+      // Generate variations off of the attributes values matrix.
+      foreach ($attribute_values_matrix as $key => $value) {
+        $variation = $this->createEntity('commerce_product_variation', [
+          'type' => 'default',
+          'sku' => $this->randomMachineName(),
+          'price' => [
+            'number' => (string) 3 * $i,
+            'currency_code' => 'USD',
+          ],
+          'attribute_color' => $color_attributes[$value[0]],
+          'attribute_size' => $size_attributes[$value[1]],
+        ]);
+        $variations[] = $variation;
+      }
+
+      $this->products[] = $this->createEntity('commerce_product', [
+        'type' => 'default',
+        'title' => $this->randomMachineName(),
+        'stores' => [$this->store],
+        'variations' => $variations,
+      ]);
+    }
+  }
+
+  /**
+   * Tests that the form IDs are unique.
+   */
+  public function testUniqueAddToCartFormIds() {
```

Comment:
I think this test method is meant to test BigPipe-powered delivery?

I replied to your question on Twitter (https://twitter.com/wimleers/status/883262976576413696), but repeating here to prevent linkrot:

> Recommendation:
> 
> 1. assert HTML placeholder
> 2. wait for the BigPipe stop signal
> 3. assert HTML placeholder has been replaced
> 
> See `\Drupal\Tests\big_pipe\FunctionalJavascript\BigPipeRegressionTest::testMultipleClosingBodies_2678662()` — waiting for the stop signal